File: Placement Sheet/LinkedList/BS/MergeSortedLinkedList.ts
class NodeLL {
    private data: number;
    private next: NodeLL | null = null;
    constructor(data?: number) {
        if (data) {
            this.data = data;
        }
    }

    getData(): number {
        return this.data;
    }

    getNext(): NodeLL | null {
        return this.next;
    }

    setNext(node: NodeLL): void {
        this.next = node;
    }
}

class MergeSortedLinkedList {
    private head: NodeLL | null = null;
    private tail: NodeLL | null = null;
    getHead(): NodeLL | null {
        return this.head;
    }

    insert(data: number): void {
        let newNode: NodeLL | null = new NodeLL(data);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
            return;
        }
        this.tail?.setNext(newNode);
        this.tail = newNode;
    }

    display(root?: NodeLL | null): void {
        if (root) {
            if (root == null) {
                console.log("List is empty...");
                return;
            }
            let currNode: NodeLL | null = root;
            let result: string = "";
            while (currNode !== null) {
                result += currNode.getData() + " --> ";
                currNode = currNode.getNext();
            }
            result += "NULL";
            console.log(result);
        } else {
            if (this.head == null) {
                console.log("List is empty...");
                return;
            }
            let currNode: NodeLL | null = this.head;
            let result: string = "";
            while (currNode !== null) {
                result += currNode.getData() + " --> ";
                currNode = currNode.getNext();
            }
            result += "NULL";
            console.log(result);
        }
    }

    mergeTwoSortedList(list1: NodeLL | null, list2: NodeLL | null): NodeLL | null {
        let dummy: NodeLL | null = new NodeLL(-1);
        let currNode: NodeLL | null | undefined = dummy;

        while (list1 !== null && list2 !== null) {
            if (list1.getData() <= list2.getData()) {
                currNode.setNext(list1);
                list1 = list1.getNext();
            } else {
                currNode?.setNext(list2);
                list2 = list2.getNext();
            }
            currNode = currNode?.getNext();
        }

        if (list1 !== null) {
            currNode?.setNext(list1);
        } else {
            currNode?.setNext(list2);
        }
        return dummy.getNext();
    }
}

let list1: MergeSortedLinkedList = new MergeSortedLinkedList();
list1.insert(1);
list1.insert(2);
list1.insert(3);
list1.insert(4);
list1.insert(5);
list1.insert(6);
console.log("List1: ");
list1.display();

let list2: MergeSortedLinkedList = new MergeSortedLinkedList();
list2.insert(7);
list2.insert(8);
list2.insert(9);
console.log("List2: ");
list2.display();

let result: MergeSortedLinkedList = new MergeSortedLinkedList();
let newHead: NodeLL | null = result.mergeTwoSortedList(list1.getHead(), list2.getHead());
result.display(newHead);

export { };

File: Placement Sheet/LinkedList/BS/MergeSortedLinkedList.java
public class MergeSortedLinkedList {
    private Node head = null;
    private Node tail = null;

    void insert(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
            return;
        }
        newNode.next = tail.next;
        tail.next = newNode;
        tail = newNode;
    }

    void display() {
        if (head == null) {
            System.out.println("List is empty...");
            return;
        }
        Node currNode = head;
        while (currNode != null) {
            System.out.print(currNode.data + " --> ");
            currNode = currNode.next;
        }
        System.out.println("NULL");
    }

    void display(Node root) {
        if (root == null) {
            System.out.println("List is empty...");
            return;
        }
        Node currNode = root;
        while (currNode != null) {
            System.out.print(currNode.data + " --> ");
            currNode = currNode.next;
        }
        System.out.println("NULL");
    }

    Node mergeTwoSortedList(Node list1, Node list2) {
        Node dummy = new Node(-1);
        Node currNode = dummy;

        while (list1 != null && list2 != null) {
            if (list1.data <= list2.data) {
                currNode.next = list1;
                list1 = list1.next;
            } else {
                currNode.next = list2;
                list2 = list2.next;
            }
            currNode = currNode.next;
        }

        if (list1 != null) {
            currNode.next = list1;
        } else {
            currNode.next = list2;
        }

        return dummy.next;
    }

    public static void main(String[] args) {
        MergeSortedLinkedList list1 = new MergeSortedLinkedList();
        list1.insert(1);
        list1.insert(2);
        list1.insert(3);
        list1.insert(4);
        list1.insert(5);
        System.out.println("List1: ");
        list1.display();

        MergeSortedLinkedList list2 = new MergeSortedLinkedList();
        list2.insert(6);
        list2.insert(7);
        list2.insert(8);
        list2.insert(9);
        System.out.println("List2: ");
        list2.display();

        MergeSortedLinkedList result = new MergeSortedLinkedList();
        Node newHead = result.mergeTwoSortedList(list1.head, list2.head);
        
        System.out.println("Merged List: ");
        result.display(newHead);
    }
}